app/src/main/kotlin/co/netguru/baby/monitor/client/feature/firebasenotification/FirebaseNotificationSender.kt

```diff
@@ -2,63 +2,94 @@ package co.netguru.baby.monitor.client.feature.firebasenotification
 import okhttp3.Response
-import org.json.JSONArray
 import org.json.JSONObject
 import timber.log.Timber
+import java.util.*
 import javax.inject.Inject
 
 class FirebaseNotificationSender @Inject constructor(
     private val dataRepository: DataRepository,
     private val httpClient: OkHttpClient,
     private val debugModule: DebugModule,
-    private val analyticsManager: AnalyticsManager
+    private val analyticsManager: AnalyticsManager,
+    private val gson: Gson
 ) {
     fun broadcastNotificationToFcm(
         title: String,
         text: String,
         notificationType: NotificationType
     ): Completable =
-        dataRepository.getAllClientData()
-            .doOnSubscribe { Timber.i("Subscribing to all client data.") }
-            .doOnNext { Timber.i("Next client data: $it.") }
-            .doOnTerminate { Timber.i("Terminated all client data.") }
-            .firstOrError()
-            .map { clients -> clients.map(ClientEntity::firebaseKey) }
-            .flatMapCompletable { firebaseTokens ->
+        dataRepository.getClientData()
+            .doOnSuccess { Timber.i("Client data: $it.") }
+            .map { client -> client.firebaseKey }
+            .flatMapCompletable { firebaseToken ->
                 analyticsManager.logEvent(Event.ParamEvent.NotificationSent(notificationType))
-                if (firebaseTokens.isNotEmpty()) {
-                    postNotificationToFcm(firebaseTokens, title, text, notificationType)
-                        .doOnSuccess { response ->
-                            debugModule.sendNotificationEvent(
-                                "Notification posted: Title: $title" +
-                                        " text: $text. Response: ${response.body?.string()}."
-                            )
-                            Timber.d("Notification posted: $response.")
-                        }
-                        .ignoreElement()
-                } else {
-                    Completable.complete()
-                }
+                postNotificationToFcm(firebaseToken, title, text, notificationType)
+                    .doOnSuccess { response ->
+                        debugModule.sendNotificationEvent(
+                            "Notification posted: Title: $title" +
+                                    " text: $text. Response: ${response.body?.string()}."
+                        )
+                        Timber.d("Notification posted: $response.")
+                    }
+                    .ignoreElement()
             }
 
     private fun postNotificationToFcm(
-        to: List<String>,
+        firebaseToken: String,
         title: String,
         text: String,
         notificationType: NotificationType
-    ): Single<Response> =
-        Single.fromCallable {
+    ): Single<Response> {
+        return checkIdCall(firebaseToken)
+            .flatMap { isAndroid ->
+                return@flatMap notificationCall(
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/1465c0ef7e5c273be21bc4e5a598037b2bec4de9/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/firebasenotification/FirebaseNotificationSender.kt#L57"></span>Expression with labels increase complexity and affect maintainability.</td>
    </tr>
  </tbody>
</table>